service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,155 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.criteria_info.CertificateCriteriaInfo;
+import com.epam.esm.criteria_info.PageInfo;
+import com.epam.esm.dto.CertificateDTO;
+import com.epam.esm.dto.CertificateUpdateDto;
+import com.epam.esm.dto.TagDTO;
+import com.epam.esm.entity.Certificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.persistence.CertificateDAO;
+import com.epam.esm.persistence.TagDAO;
+import com.epam.esm.persistence.specification.Specification;
+import com.epam.esm.persistence.specification_builder.impl.CertificateSpecificationBuilder;
+import com.epam.esm.service.CertificateService;
+import com.epam.esm.service.entitydtomapper.impl.CertificateDtoMapper;
+import com.epam.esm.service.exception.ExceptionCode;
+import com.epam.esm.service.exception.NoSuchResourceException;
+import com.epam.esm.service.validate.PaginationValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private final CertificateDAO certificateDAO;
+    private final TagDAO tagDAO;
+    private final CertificateDtoMapper certificateDtoMapper;
+    private final CertificateSpecificationBuilder specificationBuilder;
+    private final PaginationValidator paginationValidator;
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDAO certificateDAO,
+                                  CertificateDtoMapper certificateDtoMapper,
+                                  TagDAO tagDAO,
+                                  CertificateSpecificationBuilder specificationBuilder,
+                                  PaginationValidator paginationValidator) {
+        this.certificateDAO = certificateDAO;
+        this.certificateDtoMapper = certificateDtoMapper;
+        this.tagDAO = tagDAO;
+        this.specificationBuilder = specificationBuilder;
+        this.paginationValidator = paginationValidator;
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<CertificateDTO> find(PageInfo pageInfo, CertificateCriteriaInfo criteriaInfo) {
+        paginationValidator.validate(pageInfo);
+        List<Specification> specifications = specificationBuilder.build(criteriaInfo);
+        List<Certificate> certificates = certificateDAO.findAll(specifications, (int) pageInfo.getOffset(), (int) pageInfo.getLimit());
```

Comment:
Loong to int? Looks strange and dangerous. Why didn't you use int in PageInfo ?

Reply:
done